tt_metal/impl/dispatch/command_queue_interface.hpp

```diff
@@ -4,29 +4,74 @@
 
 #include "tt_metal/common/base.hpp"
 #include "tt_metal/impl/dispatch/device_command.hpp"
+#include "tt_metal/impl/dispatch/dispatch_core_manager.hpp"
 #include "tt_metal/llrt/llrt.hpp"
 #include "tt_metal/common/math.hpp"
 
 using namespace tt::tt_metal;
 
+// Starting L1 address of commands
+inline uint32_t get_command_start_l1_address(bool use_eth_l1) {
+    return (use_eth_l1 ? eth_l1_mem::address_map::ERISC_L1_UNRESERVED_BASE : L1_UNRESERVED_BASE);
+}
+
+// Where issue queue interface core pulls in data (follows command)
+inline uint32_t get_data_section_l1_address(bool use_eth_l1) {
+    uint32_t l1_base = use_eth_l1 ? eth_l1_mem::address_map::ERISC_L1_UNRESERVED_BASE : L1_UNRESERVED_BASE;
+    return l1_base + DeviceCommand::NUM_BYTES_IN_DEVICE_COMMAND;
+}
+
+// Space available in issue queue interface core pushing command data to consumer to dispatch or relay forward
+inline uint32_t get_producer_data_buffer_size(bool use_eth_l1) {
+    uint32_t l1_size = use_eth_l1 ? MEM_ETH_SIZE : MEM_L1_SIZE;
+    uint32_t l1_base = use_eth_l1 ? eth_l1_mem::address_map::ERISC_L1_UNRESERVED_BASE : L1_UNRESERVED_BASE;
+    return (l1_size - (DeviceCommand::NUM_ENTRIES_IN_DEVICE_COMMAND * sizeof(uint32_t)) - l1_base);
+}
+
+// Space available in core receiving command data to dispatch or relay forward
+inline uint32_t get_consumer_data_buffer_size(bool use_eth_l1) {
+    uint32_t num_consumer_cmd_slots = use_eth_l1 ? 1 : 2;
+    uint32_t producer_data_buffer_size = get_producer_data_buffer_size(use_eth_l1);
+    return (producer_data_buffer_size - ((num_consumer_cmd_slots - 1) * DeviceCommand::NUM_BYTES_IN_DEVICE_COMMAND)) / num_consumer_cmd_slots;
+}
+
+template <bool channel_offet>
+inline uint32_t get_cq_offset(uint16_t channel, uint8_t cq_id, uint32_t cq_size) {
+    uint32_t offset = (cq_id * cq_size);
+    if (channel_offet) {
+        offset += (DeviceCommand::MAX_HUGEPAGE_SIZE * channel);
```

Comment:
Not sure why we need to template by channel_offset. If channel is 0, we just do an addition with 0, no?

Reply:
ah there are some cases where we don't want to account for the channel offset. for eg when calling cluster apis that write to/read from sysmem we pass in address _relative_ to the channel along with the channel rather than address in sysmem from address 0

Reply:
would it make more sense to create two `get_cq_offset` functions, one with the `channel` parameter and one without?

Reply:
The current way risks the user passing a valid `channel`, but setting `channel_offset` to `false`, which renders the `channel` parameter to be not used

Reply:
good point, I can split and make one a "channel relative" offset and another is a "total offset"